angr/sim_manager.py

```diff
@@ -0,0 +1,927 @@
+                                  errored=self._errored)
+        return simgr
+
+    #
+    #   ...
+    #
+
+    def use_technique(self, tech):
+        """Use an exploration technique with this SimulationManager.
+
+        Techniques can be found in :mod:`angr.exploration_techniques`.
+
+        :param tech:    An ExplorationTechnique object that contains code to modify
+                        this SimulationManager's behavior.
+        :type tech:     ExplorationTechnique
+        :return:
+        """
+        if not isinstance(tech, ExplorationTechnique):
+            raise SimulationManagerError
+
+        # XXX: as promised
+        tech.project = self._project
+        tech.setup(self)
+
+        def _is_overriden(name):
+            return getattr(tech, name).__code__ is not getattr(ExplorationTechnique, name).__code__
+
+        overriden = filter(_is_overriden, ('step', 'filter', 'selector', 'step_state', 'successors'))
+        hooks = {name: getattr(tech, name) for name in overriden}
+        HookSet.install_hooks(self, **hooks)
+
+        self._techniques.append(tech)
+        return tech
+
+    def remove_technique(self, tech):
+        """Remove an exploration technique from a list of active techniques.
+
+        :param tech:    An ExplorationTechnique object.
+        :type tech:     ExplorationTechnique
+        :return:
+        """
+        if not isinstance(tech, ExplorationTechnique):
+            raise SimulationManagerError
+
+        def _is_overriden(name):
+            return getattr(tech, name).__code__ is not getattr(ExplorationTechnique, name).__code__
+
+        overriden = filter(_is_overriden, ('step', 'filter', 'selector', 'step_state', 'successors'))
+        hooks = {name: getattr(tech, name) for name in overriden}
+        HookSet.remove_hooks(self, **hooks)
+
+        self._techniques.remove(tech)
+        return tech
+
+    @contextlib.contextmanager
+    def adhoc_technique(self, tech=None, **more_hooks):
+        """Add a new ad-hoc technique to this manager.
+
+        The newly added technique will be removed from the simulation manager
+        upon exiting from the context.
+
+        :param tech:        A technique to add to this manager.
+        :param more_hooks:  A set of temporarly hooks to add using the TechniqueBuilder.
+        :return:            A SimulationManager instance with those techniques enabled.
+        """
+        with ImmutabilityMixin.context(self) as simgr:
+            if tech:
+                simgr.use_technique(tech)
+            other_tech = None
+            if more_hooks:
+                other_tech = TechniqueBuilder(**more_hooks)
+                simgr.use_technique(other_tech)
+            yield simgr
+            if tech:
+                simgr.remove_technique(tech)
+            if other_tech:
+                simgr.remove_technique(other_tech)
+
+    #
+    #   ...
+    #
+
+    @deprecated()
```

Comment:
your intentions are admirable, but this should probably not be deprecated. the overhead in setting up a new exploration technique is far greater than one would want to do a quick one-off "goto address".

Reply:
Agreed, I was quite excited about the refactoring. However, I think the `step_func` argument needed to be removed from the method prototype, as the step hooking could be done with `adhoc_technique`, see comment below.